go-controller/pkg/ovn/controller/unidling/unidle.go

```diff
@@ -0,0 +1,138 @@
+package unidling
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+
+	"k8s.io/klog/v2"
+)
+
+// unidlinController checks periodically the OVN events db
+// and generates a Kubernetes NeedPods events with the Service
+// associated to the VIP
+type unidlingController struct {
+	eventRecorder record.EventRecorder
+	// Map of load balancers to service namespace
+	serviceVIPToName     map[ServiceVIPKey]types.NamespacedName
+	serviceVIPToNameLock sync.Mutex
+}
+
+// NewController creates a new unidling controller
+func NewController(recorder record.EventRecorder, serviceInformer cache.SharedIndexInformer) *unidlingController {
+	uc := &unidlingController{
+		eventRecorder:    recorder,
+		serviceVIPToName: map[ServiceVIPKey]types.NamespacedName{},
+	}
+
+	// we only process events on unidling, there is no reconcilation
+	klog.Info("Setting up event handlers for services")
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: uc.onServiceAdd,
+		UpdateFunc: func(old, new interface{}) {
+			uc.onServiceDelete(old)
+			uc.onServiceAdd(new)
+		},
+		DeleteFunc: uc.onServiceDelete,
+	})
+	return uc
+}
+
+func (uc *unidlingController) onServiceAdd(obj interface{}) {
+	svc := obj.(*v1.Service)
+	if util.ServiceTypeHasClusterIP(svc) && util.IsClusterIPSet(svc) {
+		for _, ip := range util.GetClusterIPs(svc) {
+			for _, svcPort := range svc.Spec.Ports {
+				vip := util.JoinHostPortInt32(ip, svcPort.Port)
+				uc.AddServiceVIPToName(vip, svcPort.Protocol, svc.Namespace, svc.Name)
+			}
+		}
+	}
+}
+
+func (uc *unidlingController) onServiceDelete(obj interface{}) {
+	svc := obj.(*v1.Service)
+	if util.ServiceTypeHasClusterIP(svc) && util.IsClusterIPSet(svc) {
+		for _, ip := range util.GetClusterIPs(svc) {
+			for _, svcPort := range svc.Spec.Ports {
+				vip := util.JoinHostPortInt32(ip, svcPort.Port)
+				uc.DeleteServiceVIPToName(vip, svcPort.Protocol)
+			}
+		}
+	}
+}
+
+// ServiceVIPKey is used for looking up service namespace information for a
+// particular load balancer
+type ServiceVIPKey struct {
+	// Load balancer VIP in the form "ip:port"
+	vip string
+	// Protocol used by the load balancer
+	protocol v1.Protocol
+}
+
+// AddServiceVIPToName associates a k8s service name with a load balancer VIP
+func (uc *unidlingController) AddServiceVIPToName(vip string, protocol v1.Protocol, namespace, name string) {
+	uc.serviceVIPToNameLock.Lock()
+	defer uc.serviceVIPToNameLock.Unlock()
+	uc.serviceVIPToName[ServiceVIPKey{vip, protocol}] = types.NamespacedName{Namespace: namespace, Name: name}
+}
+
+// GetServiceVIPToName retrieves the associated k8s service name for a load balancer VIP
+func (uc *unidlingController) GetServiceVIPToName(vip string, protocol v1.Protocol) (types.NamespacedName, bool) {
+	uc.serviceVIPToNameLock.Lock()
```

Comment:
RLock here

Reply:
we only call `GetServiceVIPToName` in a for loop and in serial, there is no possibility to call it in parallel

Reply:
you are right. I thought we handled events in parallel threads but instead every 5 seconds we loop through all events.